### ISSUE TITLE
Oversized log file 

### ISSUE THREAD

**Issue by ZQyou** [xalt/xalt#52]
Occasionally we get an oversized XALT syslog file (> 40GB) that we cannot ingest due to out-of-memory. I notice that only few exec_path use 90% of the log file. They are tracked in few bioinformatics jobs where users run the executables in a loop. I know that we can adjust sampling rate for scalar program but we would like to track all before we decide to block unwanted ones. 

Is it possible to consolidate multiple calls of an executable in the same job (jobID) to single entry? I think this can significantly reduce the log size.

Alternatively can we add an exclude list (usually contains less than 10 lines) while doing ingest? I think this would avoid OOM issue. 

Thank you.

**Comment by rtmclay** [comment]
If the XALT syslog file is large, you can split the syslog file into a number of blocks as long as the syslog file is split on full lines and not in the middle of a line.

I believe that the OOM issue maybe related to the bug in 2.10.10 (see #48).  It should go away in the new version. 

It might be possible to add an exclude list of paths or path pattern to exclude.  That is unlikely to help with OOM but it will help in reducing the size of the database.  

However, why not just update the ignore patterns in the site config.py file and re-install xalt.  Is it that hard to update XALT on a system?

**Comment by ZQyou** [comment]
We may do split for a medium-size log file (20-30GB). For larger one, it could take a whole day to ingest and break our cron routine. Also we don't want to increase our database with those "repeated" entries because it would run out of memory of the database server when we get a usage report with SQL.

The most useful information we extract from the database is the executable usage:
```
$ xalt_usage_report 
-------------------------------------------------
XALT Software Usage from 2021-06-14 to 2021-06-20
-------------------------------------------------

Top 20 software/executables sorted by cpuhours

* Host: pitzer
* WARNING: CPUHrs is executable walltime x # cores x # threads, not actual CPU utilization
* WARNING: Runs without JobID are NOT counted

CPUHrs     NodeHrs  # Jobs  # Users  Software/Executable
------     -------  ------  -------  -------------------
193483.30  4576.56  1068    10       VASP
143112.95  3577.82  84      4        namd2
50434.84   1050.75  2831    1        nek5000
46241.61   1886.28  546     8        LAMMPS
...
```
So I was looking a possibility to consolidate the same calls in a loop/job.

We did update the ignore patterns in the site config.py when we encounter this issue. It is not hard to update XALT on our system but this update chasing is kind of endless ....

**Comment by rtmclay** [comment]
Changing the database to support combining the same executables is a major rewrite and would require a good deal of thought.  Also combining those same executables in to one entry generates many questions.  Do you combine everything named namd2 into 1 entry?  What about a.out?  This is an important topic that does not have a simple answer (at least one that I see).  Looking up an executable is expensive which means that large log files will still take a long time to ingest.  If you collect all the runs of namd2, how do you know how many users ran it?

I added the sampling feature because a 132 node cluster at TACC generated several million runs in 2 hours.  It took 4.5 days to read that data in.  There is no way to support that amount of data. That is why I implemented sampling.

I can provide a post-filter mechanism to the ingest process.  I envision a similar approach that it is part of the xalt library.  There would be another table in the Config.py file that might look something like this:

```
post_process_filter= [
    [0.0001,  r'^\/usr\/bin\/cp'],
    [0.0001,  r'^\/usr\/bin\/mv'],
    [0.0,     r'^\/usr\/bin\/gawk'],
    [0.0,    r'.*\/foobar'],
    [1.0,    r'.*']
]
```

Where the number on the left is the probability that the executable would be stored. So cp and mv would have 0.01% chance of be recorded. Where as gawk and foobar would never be recorded.  The last line would be provided automatically.

This table would be compiled into flex routine then C.  This C routine would be called by the python ingest programs (xalt_file_to_db.py and xalt_syslog_to_db.py).  Doing it this way would be much much faster than trying to use python regex.  It would having to re-build XALT just on the computer where you are ingesting *.json records.

Comments?

**Comment by rtmclay** [comment]
By the way you reported NodeHrs.  At least on our system, a user might structure a job where the ask for multiple nodes and then run multiple scalar jobs spread across all the nodes.  XALT cannot and does not know that an executable is the only task on a node or whether it is shared.

**Comment by ZQyou** [comment]
> Changing the database to support combining the same executables is a major rewrite and would require a good deal of thought. Also combining those same executables in to one entry generates many questions. Do you combine everything named namd2 into 1 entry? What about a.out? This is an important topic that does not have a simple answer (at least one that I see). Looking up an executable is expensive which means that large log files will still take a long time to ingest. If you collect all the runs of namd2, how do you know how many users ran it?

From my observations, I will say to combine the executable from the same job (job ID). It is rare that a user runs two different namd2 in the same job.

> 
> I added the sampling feature because a 132 node cluster at TACC generated several million runs in 2 hours. It took 4.5 days to read that data in. There is no way to support that amount of data. That is why I implemented sampling.

After talking to my manager,  I think we can use this sampling feature since those annoying runs are usually very short-time.

> 
> I can provide a post-filter mechanism to the ingest process. I envision a similar approach that it is part of the xalt library. There would be another table in the Config.py file that might look something like this:
> 
> ```
> post_process_filter= [
>     [0.0001,  r'^\/usr\/bin\/cp'],
>     [0.0001,  r'^\/usr\/bin\/mv'],
>     [0.0,     r'^\/usr\/bin\/gawk'],
>     [0.0,    r'.*\/foobar'],
>     [1.0,    r'.*']
> ]
> ```
> 
> Where the number on the left is the probability that the executable would be stored. So cp and mv would have 0.01% chance of be recorded. Where as gawk and foobar would never be recorded. The last line would be provided automatically.
> 
> This table would be compiled into flex routine then C. This C routine would be called by the python ingest programs (xalt_file_to_db.py and xalt_syslog_to_db.py). Doing it this way would be much much faster than trying to use python regex. It would having to re-build XALT just on the computer where you are ingesting *.json records.
> 
> Comments?

This sounds interesting. Would this improve ingest time for a large log file? I am still working on reducing the log files by removing  unwanted entries.

**Comment by rtmclay** [comment]
I have not written the feature to implement a post_process_filter but it is very likely that supporting this filter would be much much faster than trying to interact with the database.   I wanted you to understand that the filter would be converted into a compiled C function called from python.  This would require an install of XALT on the computer that is ingesting the *.json records into the data base.  It would not require re-installing XALT on the compute or login nodes.

There might be a way to merge json records into the data base but that is expensive when trying to ingest the data.

Say you have a high throughput job.  Where a single node is running the same executable say 40 times with different input.  It might be possible to combine those records, but there would have to a daemon running on a compute node to handle that decision process.  There currently is no daemon installed and I don't know what that means.   

Before I implemented sampling, I tried to do filtering based on job_id and that was too expensive.

So are you or any other sites interested in having the post_process_filter support added?  I'll do the work.
